casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala

```diff
@@ -178,37 +170,56 @@ class Initializing[F[_]
     type SortedBlocks = SortedMap[Long, Set[BlockHash]]
     type RecParams    = (Seq[BlockHash], Set[BlockHash], SortedBlocks)
 
-    def loopDependencies(params: RecParams): F[Either[RecParams, SortedBlocks]] = {
-      val (hashes, skipped, sorted) = params
+    def loopDependencies(
+        params: RecParams
+    ): F[Either[RecParams, SortedMap[Long, Set[BlockHash]]]] = {
+      val (hashes, visited, sortedByHeight) = params
       hashes match {
-        case Nil => sorted.asRight[RecParams].pure[F]
+        case Nil => sortedByHeight.asRight[RecParams].pure[F]
         case head +: tail =>
           for {
-            block       <- BlockStore[F].getUnsafe(head)
-            depsAll     = ProtoUtil.dependenciesHashesOf(block)
-            depsNext    <- depsAll.filter(skipped).filterA(BlockStore[F].contains)
-            depsSkipped = depsAll.filter(depsNext.contains)
+            block <- BlockStore[F].getUnsafe(head)
+            _     <- Log[F].info(s"placing ${PrettyPrinter.buildString(block, short = true)}")
+
+            // Block's all dependencies
+            depsAll = ProtoUtil.dependenciesHashesOf(block).filterNot(visited)
+
+            // Dependencies to add to DAG (without skipped or not in the store)
+            depsInStore <- depsAll.filterA(BlockStore[F].contains)
+
+            // Already collected hashes for block height
             blockNumber = block.body.state.blockNumber
-            nrHashes    = sorted.getOrElse(blockNumber, Set())
+            nrHashes    = sortedByHeight.getOrElse(blockNumber, Set())
+
             // Data for continue recursion
-            newRest    = tail ++ depsNext
-            newSkipped = skipped ++ depsSkipped
-            newSorted  = sorted.updated(blockNumber, nrHashes + block.blockHash)
-          } yield (newRest, newSkipped, newSorted).asLeft
+            newRest    = tail ++ depsInStore
+            newVisited = visited ++ depsAll
+            newSorted  = sortedByHeight.updated(blockNumber, nrHashes + block.blockHash)
+          } yield (newRest, newVisited, newSorted).asLeft
       }
     }
 
     val emptySorted = SortedMap[Long, Set[BlockHash]]()
     for {
+      _ <- Log[F].info(s"Adding blocks for approved state to DAG.")
       sortedHashes <- (Seq(startBlock.blockHash), Set[BlockHash](), emptySorted)
                        .tailRecM(loopDependencies)
       // Add sorted DAG in reverse order (from approved block)
       _ <- sortedHashes.flatMap(_._2).toList.reverse.traverse_ { hash =>
             for {
               block <- BlockStore[F].getUnsafe(hash)
-              _     <- BlockDagStorage[F].insert(block, invalid = false)
+              // if sender has stake 0 in approved block, this means that sender has been slashed and block is invalid
+              isInvalid = startBlock.body.state.bonds.exists(bond => {
+                bond.validator == block.sender && bond.stake == 0L
+              })
```

Comment:
This seems to mark all blocks from slashed validator as invalid although only latest message should be invalid. :thinking:

Reply:
Fixed.